pkg/accounting/accounting.go

```diff
@@ -243,103 +278,54 @@ func (a *Accounting) settle(ctx context.Context, peer swarm.Address, balance *ac
+	// compute the debt including debt created by incoming payments
+	compensatedBalance, err := a.CompensatedBalance(peer)
 	if err != nil {
-		return fmt.Errorf("failed to get surplus balance: %w", err)
+		return err
 	}
-	if surplusBalance.Cmp(big.NewInt(0)) > 0 {
 
-		// get new surplus balance after deduct
-		newSurplusBalance := new(big.Int).Sub(surplusBalance, cost)
+	paymentAmount := new(big.Int).Neg(compensatedBalance)
 
-		// if nothing left for debiting, store new surplus balance and return from debit
-		if newSurplusBalance.Cmp(big.NewInt(0)) >= 0 {
-			a.logger.Tracef("surplus debiting peer %v with value %d, new surplus balance is %d", peer, price, newSurplusBalance)
+	// Don't do anything if there is no actual debt or no time passed since last refreshment attempt
+	// This might be the case if the peer owes us and the total reserve for a peer exceeds the payment treshold.
+	if paymentAmount.Cmp(big.NewInt(0)) > 0 && timeElapsed > 0 {
+		shadowBalance, err := a.shadowBalance(peer)
+		if err != nil {
+			return err
+		}
 
-			err = a.store.Put(peerSurplusBalanceKey(peer), newSurplusBalance)
-			if err != nil {
-				return fmt.Errorf("failed to persist surplus balance: %w", err)
-			}
-			// count debit operations, terminate early
-			a.metrics.TotalDebitedAmount.Add(float64(price))
-			a.metrics.DebitEventsCount.Inc()
-			return nil
+		acceptedAmount, timestamp, err := a.refreshFunction(context.Background(), peer, paymentAmount, shadowBalance)
+		if err != nil {
+			return fmt.Errorf("refresh failure: %w", err)
 		}
 
-		// if surplus balance didn't cover full transaction, let's continue with leftover part as cost
-		debitIncrease := new(big.Int).Sub(new(big.Int).SetUint64(price), surplusBalance)
+		balance.refreshTimestamp = timestamp
 
-		// conversion to uint64 is safe because we know the relationship between the values by now, but let's make a sanity check
-		if debitIncrease.Cmp(big.NewInt(0)) <= 0 {
-			return fmt.Errorf("sanity check failed for partial debit after surplus balance drawn")
-		}
-		cost.Set(debitIncrease)
+		oldBalance = new(big.Int).Add(oldBalance, acceptedAmount)
 
-		// if we still have something to debit, than have run out of surplus balance,
-		// let's store 0 as surplus balance
-		a.logger.Tracef("surplus debiting peer %v with value %d, new surplus balance is 0", peer, debitIncrease)
+		a.logger.Tracef("registering refreshment sent to peer %v with amount %d, new balance is %d", peer, acceptedAmount, oldBalance)
 
-		err = a.store.Put(peerSurplusBalanceKey(peer), big.NewInt(0))
+		err = a.store.Put(peerBalanceKey(peer), oldBalance)
 		if err != nil {
-			return fmt.Errorf("failed to persist surplus balance: %w", err)
+			return fmt.Errorf("settle: failed to persist balance: %w", err)
 		}
-
 	}
 
-	currentBalance, err := a.Balance(peer)
-	if err != nil {
-		if !errors.Is(err, ErrPeerNoBalance) {
-			return fmt.Errorf("failed to load balance: %w", err)
+	if a.payFunction != nil {
+		if !balance.paymentOngoing {
```

Comment:
Can be in line with `314` to reduce nesting.

Reply:
done